### PR TITLE
RLP-956: integrate new rif-comms version

### PR DESCRIPTION
**Pull request by mortelli**
This PR solves issue [RLP-956](https://jirainfuy.atlassian.net/browse/RLP-956), which is about integrating the new version of the RIF Comms Pub-sub node into Lumino.

## Overview
Besides adding the new proto files, the RIF Comms client code had its signatures updated.

With the aim of validating these new functionalities, the integration test suite for RIF Comms transport was extended.

### RIF Comms client
- the `connect` method does not have a `return` statement anymore; it was unneeded and unused.
- the `_get_peer_id` method was added back to the client. it is currently only used to verify correctness in the integration tests.
- `subscribe_to` and `unsubscribe_from` have had their signatures updated to match the latest pubsub implementation.
- `is_subscribed_to` is now an internal function. it is currently only used to verify correctness in the integration tests; we no longer check for subscription in order to send a raiden message.
- `disconnect` no longer calls `EndCommunication`, as the latest version for this call does nothing. see pending issue [here](https://jirainfuy.atlassian.net/browse/RLP-925)

### RIF Comms integration tests
- 3 separate tests for the client `connect` function were added.
  - since we don't want automatic connection for the comms nodes in these tests, the existing fixture was not used here.
  - an `auto_connect` field was added to the `Config` and `Node` test classes in order to re-use them without automatic connection. its default value is `True` so that this is all compatible with the existing `comms_nodes` fixture.
  - the aforementioned tests are:
      1. boot up a comms node and create a client with an RSK address. first, check that the address cannot be found in the comms node. then, connect and check that it can. this test currently **fails** because of an incorrect exception.
      2. boot up a comms node and create a client with an RSK address. connect to the node and get the correspondent peer id. then, repeat the connection call, and check that the peer id remains the same.
      3. boot up 2 comms nodes and connect separate clients with different RSK addresses for each of them. then, try to locate the peer ID from each client based on the other client's address.
- since the client method `_is_subscribed_to` is (now) internal, its separate tests were removed. however, the logic in these tests was merged into the related `subscribe_to` tests, which now use this internal function as a means of verification.
    - `test_is_subscribed_to_invalid` logic is now in `test_subscribe_to_invalid`
    - `test_is_subscribed_to_self` logic is now in `test_subscribe_to_self`
    - `test_is_subscribed_to_peers` logic is now in `test_subscribe_to_peers`
-  a new `test_send_message_shutdown` test was added, which verifies no exception is raised when sending a message to an address which was registered in a comms node which is shut down immediately before sending the message.
- `test_send_message_subscription` now sends a message before subscribing to a topic, as well as after doing so. both are expected to work based on the latest pubsub implementation.

#### Untested scenarios
1. full coverage for exceptions raised by the comms node or through the comms client. this should be completed through [RLP-900](https://jirainfuy.atlassian.net/browse/RLP-900)
2. multiple connections on the same node. this should be solved through [RLP-969](https://jirainfuy.atlassian.net/browse/RLP-969) and #185, and eventually [RLP-955](https://jirainfuy.atlassian.net/browse/RLP-955)
3. we are still missing a test case for a connection failure, see [RLP-974](https://jirainfuy.atlassian.net/browse/RLP-974)
4. attempting to connect with the same RSK address on different nodes is not currently being tested. i believe this should be tackled when implementing [RLP-943](https://jirainfuy.atlassian.net/browse/RLP-943)
5. tests for `disconnect`. these should be solved through [RLP-925](https://jirainfuy.atlassian.net/browse/RLP-925) and [RLP-926](https://jirainfuy.atlassian.net/browse/RLP-926)

## TODO
- [x] verify `pytest` results
- [x] verify manual test cases

## Solved issues
Besides [RLP-956](https://jirainfuy.atlassian.net/browse/RLP-956), this PR will also close:
1. [RLP-897: fix connect](https://jirainfuy.atlassian.net/browse/RLP-897)
2. [RLP-957: test connect](https://jirainfuy.atlassian.net/browse/RLP-957)

## Pending related issues
1. [RLP-900: handling exceptions](https://jirainfuy.atlassian.net/browse/RLP-900)
2. [RLP-925: fix disconnect](https://jirainfuy.atlassian.net/browse/RLP-925)
3. [RLP-926: test disconnect](https://jirainfuy.atlassian.net/browse/RLP-926)
4. [RLP-936: rif comms test flakes](https://jirainfuy.atlassian.net/browse/RLP-936), which is blocked by [RLP-941](https://jirainfuy.atlassian.net/browse/RLP-941)
5. [RLP-941: rif comms ping/healthcheck call](https://jirainfuy.atlassian.net/browse/RLP-941)
6. [RLP-937: rif comms test coverage](https://jirainfuy.atlassian.net/browse/RLP-937); this one is partially resolved in this PR but not fully
7. [RLP-943: rif comms handshake](https://jirainfuy.atlassian.net/browse/RLP-943)
8. [RLP-955: rif comms edge case tests](https://jirainfuy.atlassian.net/browse/RLP-955)
9. [RLP-969: rif comms basic test cases](https://jirainfuy.atlassian.net/browse/RLP-969)
10. [RLP-971: send message to invalid address](https://jirainfuy.atlassian.net/browse/RLP-971)
11. [RLP-972: locate invalid peer ID error](https://jirainfuy.atlassian.net/browse/RLP-972)
12. [RLP-973: invalid subscription error](https://jirainfuy.atlassian.net/browse/RLP-973)
13. [RLP-974: failed connect](https://jirainfuy.atlassian.net/browse/RLP-974)
14. [RLP-975: invalid unsubscription error](https://jirainfuy.atlassian.net/browse/RLP-975)

### CODE REVIEW

File: raiden/api/python.py [hunk 1]
```diff
@@ -73,8 +73,7 @@
     PaymentMappingState
```
Comment (mortelli) [inline raiden/api/python.py:73]: changes to this file are unrelated to the issue at hand, they're just for the sake of tidiness

File: transport/rif_comms/client.py [hunk 2]
```diff
@@ -27,15 +27,23 @@ def __init__(self, rsk_address: Address, grpc_api_endpoint: str):
...
         Connects to RIF Communications Node.
         Invokes ConnectToCommunicationsNode GRPC API endpoint.
         Adds the client RSK address under the RIF Communications node peer ID.
-        :return: notification stream
         """
-        # FIXME: communications should not need variable assignment
-        return self.stub.ConnectToCommunicationsNode(self.rsk_address)
+        self.stub.ConnectToCommunicationsNode(self.rsk_address)
+
+    def _get_peer_id(self, rsk_address: Address) -> str:
```
Comment (mortelli) [inline transport/rif_comms/client.py:38]: only used for integration tests now

File: transport/rif_comms/client.py [hunk 3]
```diff
@@ -49,14 +57,17 @@ def subscribe_to(self, rsk_address: Address) -> (str, Notification):
...
+                subscriber=self.rsk_address
+            )
         )
         for response in topic:
             topic_id = response.channelPeerJoined.peerId
             break
         return topic_id, topic
 
-    def is_subscribed_to(self, rsk_address: Address) -> bool:
+    def _is_subscribed_to(self, rsk_address: Address) -> bool:
```
Comment (mortelli) [inline transport/rif_comms/client.py:70]: only used for integration tests now

File: transport/rif_comms/client.py [hunk 4]
```diff
@@ -90,14 +101,17 @@ def unsubscribe_from(self, rsk_address: Address):
...
 
     def disconnect(self):
         """
          Invokes the EndCommunication GRPC API endpoint.
          Disconnects from RIF Communications Node. Closes grpc connection
         """
-        # FIXME: ending communications should not need variable assignment
-        disconnection = self.stub.EndCommunication(Void())
+        # FIXME: EndCommunication pending implementation
+        # self.stub.EndCommunication(Void())
```
Comment (mortelli) [inline transport/rif_comms/client.py:116]: see [RLP-925](https://jirainfuy.atlassian.net/browse/RLP-925)

File: raiden/tests/integration/network/transport/rif_comms/process.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def start(env_name: str) -> Popen:
         )
 
         # FIXME: we need some sort of ping call, this might sometimes not be enough
-        time.sleep(5)  # hack to get around calling the comms node before it is ready
+        time.sleep(7)  # hack to get around calling the comms node before it is ready
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/process.py:36]: reduces the flakiness likelihood when running the test suite.

more details in [RLP-936](https://jirainfuy.atlassian.net/browse/RLP-936)

File: raiden/tests/integration/network/transport/rif_comms/test_client.py [hunk 6]
```diff
@@ -23,23 +23,102 @@ def comms_nodes(amount_of_nodes) -> {int, CommsNode}:
         node.stop()
 
 
+@pytest.mark.xfail(reason="wrong exception message from comms node")
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/test_client.py:26]: see [RLP-972](https://jirainfuy.atlassian.net/browse/RLP-972)

File: raiden/tests/integration/network/transport/rif_comms/test_client.py [hunk 7]
```diff
@@ -246,33 +341,8 @@ def test_unsubscribe_from_peers(comms_nodes):
...
-
-    for resp in sub_1_to_2:
-        sender = get_sender_from_notification(resp)
-        assert sender == address_1
-        break  # only 1 message is expected
-
-    for resp in sub_2_to_1:
-        sender = get_sender_from_notification(resp)
-        assert sender == address_2
-        break  # only 1 message is expected
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/test_client.py:278]: this was moved further up, not removed

File: raiden/tests/integration/network/transport/rif_comms/test_client.py [hunk 8]
```diff
@@ -143,70 +252,56 @@ def test_send_message_subscription(comms_nodes):
...
+    payload = "polo"
+    client_1.send_message(payload, address_2)
     for resp in sub:
         received_message = notification_to_payload(resp)
         assert received_message == payload
         break  # only 1 message is expected
 
 
-@pytest.mark.parametrize("amount_of_nodes", [2])
-def test_send_message_peers(comms_nodes):
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/test_client.py:181]: this one was also moved further up

File: raiden/tests/integration/network/transport/rif_comms/test_client.py [hunk 9]
```diff
@@ -143,70 +252,56 @@ def test_send_message_subscription(comms_nodes):
...
     for resp in sub:
         received_message = notification_to_payload(resp)
         assert received_message == payload
         break  # only 1 message is expected
 
+    # now subscribe from node 1 to 2
+    client_1.subscribe_to(address_2)
 
-@pytest.mark.parametrize("amount_of_nodes", [1])
-def test_send_message_self(comms_nodes):
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/test_client.py:166]: also moved

File: raiden/tests/integration/network/transport/rif_comms/test_client.py [hunk 10]
```diff
@@ -143,70 +252,56 @@ def test_send_message_subscription(comms_nodes):
     # node 2 must listen to its own topic
     _, sub = client_2.subscribe_to(address_2)
 
-    # send message without subscription
+    # send message and receive without subscription
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/test_client.py:254]: this is no longer expected to fail based on the latest pub-sub code

File: raiden/tests/integration/network/transport/rif_comms/test_client.py [hunk 11]
```diff
@@ -63,75 +153,93 @@ def test_subscribe_to_repeated(comms_nodes):
...
+def test_send_message_invalid(comms_nodes):
     client, address = comms_nodes[1].client, comms_nodes[1].address
 
-    # check subscription to non-subscribed address
-    assert client.is_subscribed_to(address) is False
-
-    # check subscription to unregistered address
-    assert client.is_subscribed_to(generate_address()) is False
+    # send message to unregistered peer
+    # FIXME: hangs
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/test_client.py:173]: see [RLP-971](https://jirainfuy.atlassian.net/browse/RLP-971)

File: raiden/tests/integration/network/transport/rif_comms/test_client.py [hunk 12]
```diff
@@ -23,23 +23,101 @@ def comms_nodes(amount_of_nodes) -> {int, CommsNode}:
...
+
+    # no subscriptions should be present
+    assert client._is_subscribed_to(unregistered_address) is False
+
     # subscribe to unregistered address
-    topic_id, _ = client.subscribe_to(generate_address())
+    topic_id, _ = client.subscribe_to(unregistered_address)
 
+    # check subscription again
     # FIXME: this should probably throw an exception rather than an empty result
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/test_client.py:105]: see [RLP-973](https://jirainfuy.atlassian.net/browse/RLP-973)

File: raiden/tests/integration/network/transport/rif_comms/test_client.py [hunk 13]
```diff
@@ -143,70 +251,56 @@ def test_send_message_subscription(comms_nodes):
...
+    finally:
+        for node in nodes:
+            # because node 2 can be already stopped, we cannot call stop() indiscriminately
+            if not node._process.poll():  # if True, node is still running
+                node.stop()
 
 
 @pytest.mark.parametrize("amount_of_nodes", [1])
-@pytest.mark.xfail(reason="incorrect exception from comms")
+@pytest.mark.xfail(reason="exceptions are not raised from comms node")
```
Comment (mortelli) [inline raiden/tests/integration/network/transport/rif_comms/test_client.py:303]: see [RLP-975](https://jirainfuy.atlassian.net/browse/RLP-975)

**Comment by mortelli** [comment]
@marcosmartinez7 @JONAF2103 

please take a look at `raiden/tests/integration/network/transport/rif_comms/test_client.py` in your IDE; the diff in this PR is too messy to properly review the changes in this module.

thanks

**Review by JONAF2103 — APPROVED** [review]
LGTM

**Comment by marcosmartinez7** [comment]
Great work. Lgtm in general. 

I have reviewed the test cases mainly, and i have merge them into the PR that builds on this https://github.com/rsksmart/lumino/pull/192

Some things to keep in mind:

1. i have manually merged the rest of the test cases but now using the multi scenario cluster
2. i added the new test cases (connect repeat, shutdown) to the  PR #192 

From the 2 points above, i think the PR #192 has all the changes of this PR but we can compare it later. 

I have the manual test pending too.

**Review by marcosmartinez7 — APPROVED** [review]
